smithy-utils/src/main/java/software/amazon/smithy/utils/CodeFormatter.java

```diff
@@ -15,369 +15,331 @@
 
-    private void parseArgumentWrapper(CodeWriter writer, State state) {
-        if (state.eof()) {
-            throw error(writer, "Invalid format string: " + state);
+        // Writes literal segments of the input string.
+        static Operation stringSlice(String source, int start, int end) {
+            return (sink, writer, column) -> sink.append(source, start, end);
         }
 
-        char expressionStart = writer.getExpressionStart();
-        char c = state.c();
-        if (c == expressionStart) {
-            // $$ -> $
-            state.append(expressionStart);
-            state.next();
-        } else if (c == '{') {
-            parseBracedArgument(writer, state);
-        } else {
-            parseArgument(writer, state, -1);
+        // An operation that writes a resolved variable to the writer (e.g., positional arguments).
+        static Operation staticValue(String value) {
+            return (sink, writer, column) -> sink.append(value);
         }
-    }
-
-    private void parseBracedArgument(CodeWriter writer, State state) {
-        int startingBraceColumn = state.column;
-        state.next(); // Skip "{"
-        parseArgument(writer, state, startingBraceColumn);
 
-        if (state.eof() || state.c() != '}') {
-            throw error(writer, "Unclosed expression argument: " + state);
+        // Expands inline sections.
+        static Operation inlineSection(String sectionName, Operation delegate) {
+            return (sink, writer, column) -> {
+                StringBuilder buffer = new StringBuilder();
+                delegate.apply(buffer, writer, column);
+                sink.append(writer.expandSection(sectionName, buffer.toString(), writer::writeWithNoFormatting));
+            };
         }
 
-        state.next(); // Skip "}"
-    }
-
-    private void parseArgument(CodeWriter writer, State state, int startingBraceColumn) {
-        if (state.eof()) {
-            throw error(writer, "Invalid format string: " + state);
-        }
-
-        char c = state.c();
-        if (Character.isLowerCase(c)) {
-            parseNamedArgument(writer, state, startingBraceColumn);
-        } else if (Character.isDigit(c)) {
-            parsePositionalArgument(writer, state, startingBraceColumn);
-        } else {
-            parseRelativeArgument(writer, state, startingBraceColumn);
+        // Used for "|". Wraps another operation and ensures newlines are properly indented.
+        static Operation block(Operation delegate) {
+            return (sink, writer, column) -> delegate.apply(new BlockAppender(sink, column), writer, column);
         }
     }
 
-    private void parseNamedArgument(CodeWriter writer, State state, int startingBraceColumn) {
-        // Expand a named context value: "$" key ":" identifier
-        String name = parseNameUntil(writer, state, ':');
-        state.next();
+    private static final class BlockAppender extends DecoratedAppendable {
```

Comment:
This doesn't seem to account for the indentation settings on the codewriter. Go, for instance, uses tabs instead of spaces. So for Go it would produce:

```go
func main() {
 fmt.Println("Hello world")
}
```

But also, and maybe I'm missing something here, it looks like the column is just column and not whitespace so it would end up being more like:

```go
func main() {
             fmt.Println("Hello world")
}
```

Reply:
That's exactly how it works today too, before this refactor. If you use `|` for block formatting, it takes into account the number of spaces before that interpolation, dynamically accounting for other things that might have been interpolated before it. In order to properly indent with tabs and potentially a mix of tabs and spaces, we'd need to introduce static and dynamic formatting for blocks.

* Static would be used if the characters on a line before interpolation are all spaces and tabs, then that's the text used at each new line.
* Dynamic would be used when the characters before the text are not all spaces and tabs. This can only use spaces since we can't base the exact alignment on anything else. Relying on the configured indent text of CodeWriter doesn't work since block formatting is supposed to align perfectly with where the interpolation is defined.

Reply:
I updated the PR to use this approach.